doc/gearbox/gearbox_mdio-HLD.md

```diff
@@ -0,0 +1,298 @@
+	- gbsyncd will have dependencies on syncd to create the Unix IPC socket. 
+	- Syncd will have new classes for the MDIO IPC server and for the parser of gearbox configuration.
+	- There is a new MDIO IPC client library is added.
+	- External PHY firmware download time always has performance requirements/impact.
+	- For debugging, beside the syslog logging, the Unix socket IPC mechanism can be simulated by socat.
+	- The current change is only limited to the Broadcom switch NPU platform, but should work for other NPU platforms.
+	- SAI switch api has added the mido clause 22 access functions.
+	- The SaiInterface class has added the mdio clause 45 read/write and mdio clause 22 read/write virtual functions.
+	- The VendoSai class overrides the mdio clause 45 read/write and mdio clause 22 read/write virtual functions.
+	- New keys are added to the gearbox configuration file "gearbox\_config.json".
+	- During the warmboot, the creation of the Unix IPC socket and the IPC connection is the same as of coldboot.
+	- The platform module software should not reset the external PHY during warmboot.
+
+When the VendorSai class constructor with PHY id argument is used in syncd.cpp, we can assume a simple case where syncd context number is the same as the PHY id.
+
+	int syncd_main(int argc, char **argv)
+	{
+	    ...
+	    if (commandLineOptions->m_globalContext != 0) {   
+	        auto vendorSai = std::make_shared<VendorSai>(
+	        (int)commandLineOptions->m_globalContext,
+	        (int)commandLineOptions->m_globalContext);
+	        ...
+	    } else {   
+	        auto vendorSai = std::make_shared<VendorSai>();
+	        ...
+	    }
+	    ...
+	}
+
+For the case that multiple PHY ids aggregate in a single syncd context, it is a new separate topic. We don't discuss the topic in this HLD.
+
+The VendorSai overloading constructor will use the functions gbsyncd\_get\_pai\_lib\_name(phy\_id),  gbsyncd\_get\_phy\_access\_lib\_name(phy\_id) and gbsyncd\_get\_mdio\_cl22\_only(phy\_id) to get the JSON key values for "lib\_name", "phy\_access\_lib\_name" and "mdio\_cl22\_only" in the configuration file "gearbox\_config.json".
+
+	#define MDIO_READ            "mdio_read"
+	#define MDIO_WRITE           "mdio_write"
+	#define MDIO_READ_CLAUSE22   "mdio_read_cl22"
+	#define MDIO_WRITE_CLAUSE22  "mdio_write_cl22"
+
+	VendorSai::VendorSai(int phy_id, int context)
+	{
+	    ...
+	    m_context = context;
+	    m_sai_dll_handle = dlopen (gbsyncd_get_pai_lib_name().c_str(), RTLD_LAZY);
+	    *(void**)(&dll_sai_api_initialize) = dlsym(m_sai_dll_handle, "sai_api_initialize");
+	    *(void**)(&dll_sai_api_uninitialize) = dlsym(m_sai_dll_handle, "sai_api_uninitialize");
+	    *(void**)(&dll_sai_api_query) = dlsym(m_sai_dll_handle, "sai_api_query");
+	    ...
+	    m_access_lib_handle = dlopen (gbsyncd_get_phy_access_lib_name().c_str(), RTLD_LAZY);
+	    m_mdio_cl22_only = gbsyncd_get_mdio_cl22_only();
+	    ...
```

Comment:
Using Broncos/Broadcom PAI as an example, the MDIO IPC client is part of the PAI. We think the MDIO IPC client and PAI should be separate libraries. With a separate PAI library which does not include the MDIO access library, we need another mechanism to specify what kind of MDIO access method to use, e.g. IPC vs. SYSFS. dlopen()/dlsym() is one of the ways to provide the mechanism.

Reply:
Like MDIO IPC server in syncd, MDIO IPC client could be a part of gbsyncd. This is common and mdio read/write is provided to PAI with switch attr SAI_SWITCH_ATTR_REGISTER_READ/SAI_SWITCH_ATTR_REGISTER_WRITE.

PAI itself knows MIDIO access method. If it is SYSFS, it can implement its mdio read/write with SYSFS and ignore attr SAI_SWITCH_ATTR_REGISTER_READ/SAI_SWITCH_ATTR_REGISTER_WRITE.

Further more, we could create a common IPC channel between syncd and gbsyncd. MDIO access is a service on top of it. In the future new service could be implemented on it.

Reply:
The MDIO SYSFS interface may depend on FPGA/CPLD based MDIO kernel driver implementation. There will be many different hardware implementation of the FPGA/CPLD hence the SYSFS interface could differ from vendor to vendor. The easiest way to add new MDIO access function is to add the function through library.
The current common IPC mechanism in SONiC is RedisDB. We have tried to use RedisDB, the performance is very bad. When any common IPC channel is used, it puts a limit on using a specific thread to process MDIO only, which affects the performance of firmward download.

Reply:
MDIO SYSFS access is one part of PAI lib, which is vendor specific.

For common IPC mechanism, I means your IPC client/server for MDIO access.

Reply:
The IPC channel is dedicated to MDIO access. If the channel is shared with other IPC tasks, we could see performance issue.

Reply:
We will change to pass above values to VendorSai constructor.